### PR TITLE
Use COPY instead of ADD

### PR DESCRIPTION
**Pull request by mkam**
I forgot to use COPY instead of ADD. ADD is more likely to cause cache invalidations, so that's why Docker kept having to reinstall the python packages.

https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

Connects https://github.com/rcbops/u-suk-dev/issues/1539

### CODE REVIEW

**Comment by mkam** [comment]
recheck_lint

**Comment by mkam** [comment]
recheck_lint

**Comment by mkam** [comment]
Our version of Docker is still inspecting the modified time to invalidate cache on ADD/COPY, so this isn't currently working in Jenkins. We may need to upgrade it.

https://github.com/docker/docker/issues/9391#issuecomment-272308804

**Comment by hughsaunders** [comment]
Seems to fixed in docker 1.8, but 1.7 (currently installed) is the latest version availabe in epel 6 (the slave is running centos 6.8). https://dl.fedoraproject.org/pub/epel/6/x86_64/

**Comment by hughsaunders** [comment]
Heh so an [SO answer](http://stackoverflow.com/questions/26531108/docker-add-cache-when-git-checkout-same-file/26612694#26612694) suggests using  `touch -t` with a fixed value so that the file content and metadata matches that expected by docker. Worth a try..

**Comment by hughsaunders** [comment]
recheck_lint

**Comment by hughsaunders** [comment]
recheck_lint

**Comment by hughsaunders** [comment]
recheck_lint

**Comment by hughsaunders** [comment]
It works :)
```
[Pipeline] sh
[RPC-Gating-Lint] Running shell script
+ touch -t 201704100000 constraints.txt requirements.txt
[Pipeline] sh
[RPC-Gating-Lint] Running shell script
+ docker build -t jenkins-rpc-gating-lint-166 .
Sending build context to Docker daemon 546.3 kB
Sending build context to Docker daemon 546.3 kB
Sending build context to Docker daemon 
Step 0 : FROM ubuntu
 ---> 7f06d5cab2df
Step 1 : RUN apt-get update && apt-get install -y groovy2 python-pip build-essential python-dev libssl-dev
 ---> Using cache
 ---> 350a0fbd773f
Step 2 : RUN apt-get install -y libffi-dev
 ---> Using cache
 ---> 8557487b8486
Step 3 : RUN apt-get install -y sudo
 ---> Using cache
 ---> 869036612b01
Step 4 : COPY requirements.txt /requirements.txt
 ---> Using cache
 ---> a16e21efc870
Step 5 : COPY constraints.txt /constraints.txt
 ---> Using cache
 ---> dc2c37a6ead5
Step 6 : RUN pip install -c /constraints.txt -r /requirements.txt
 ---> Using cache
 ---> 601f2f25b5aa
Step 7 : RUN useradd jenkins --shell /bin/bash --create-home --uid 500
 ---> Using cache
 ---> ecd3108a92f1
Step 8 : RUN echo "jenkins ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
 ---> Using cache
 ---> fe6d74eb3c57
Successfully built fe6d74eb3c57
```
https://rpc.jenkins.cit.rackspace.net/job/RPC-Gating-Lint/166/console

File: rpc_jobs/rpc_gating_lint.yml [hunk 1]
```diff
@@ -18,6 +19,11 @@
         deleteDir()
         stage("Prepare"){
           git url: env.ghprbAuthorRepoGitUrl, branch: ghprbSourceBranch
+          /* Set mtime to a constant value as git doesn't track mtimes but
```
Comment (git-harry) [inline rpc_jobs/rpc_gating_lint.yml]: I think it would be nice to create a function for this so we're not copying the same thing around.

File: rpc_jobs/rpc_gating_lint.yml [hunk 2]
```diff
@@ -18,6 +19,11 @@
...
+           * docker 1.7 does, this causes cache invalidation when files are
+           * added
+           */
+          sh "touch -t 201704100000 *.txt"
```
Comment (mkam) [inline rpc_jobs/rpc_gating_lint.yml]: If we always set the mtime for the files, then we'll run into issues if we do actually update the files.

**Comment by hughsaunders** [comment]
recheck_lint

**Comment by hughsaunders** [comment]
recheck_lint

File: rpc_jobs/rpc_gating_lint.yml (hunk 1 above)
Comment (hughsaunders) [inline rpc_jobs/rpc_gating_lint.yml]: Done

File: rpc_jobs/rpc_gating_lint.yml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 - job:
     name: 'RPC-Gating-Lint'
     project-type: workflow
+    concurrent: true
```
Comment (git-harry) [inline rpc_jobs/rpc_gating_lint.yml]: Why was this added?

Comment (mattt416) [inline rpc_jobs/rpc_gating_lint.yml]: concurrent defaults to `false`, and there is no reason why this job cannot run concurrently.
